src/Stmt.h

```diff
@@ -66,10 +67,13 @@ class Stmt : public Obj {
 
 	virtual TraversalCode Traverse(TraversalCallback* cb) const = 0;
 
+#include "script_opt/StmtOpt-Public.h"
```

Comment:
Not a fan of this (and similar pieces elsewhere). Just flagging for now, need to think about other options.

Reply:
FYI, I tried going the multiple inheritance route that you'd suggested.  There's a comment at the beginning of `script_opt/ExprOpt-Public.h`: https://github.com/zeek/zeek/blob/96848687c98eaaa4908834dee26737ec6f3d5e53/src/script_opt/ExprOpt-Public.h#L5 of why I concluded it would be quite a bit uglier still.  The core problem is that the `Stmt` and `Expr` subclasses need to override the optimization-related virtual methods, which means the change is actually more invasive.

I also gave thought to whether this could be done using accessors/mutators, but ultimately that's going to require a lot of massive switch statements, which seems to me antithetical to object-oriented programming.

Reply:
I may have lost track of some larger goals / problem-definitions, so sorry for potentially adding noise/confusion but I'm not sure it's worthwhile to jump through the script-optimization demarcation hoops for this case:

* Even though it's practically expected to only be used by script-opt logic, AST-duplication is conceptually independent and that's what allows it to be a candidate for independent merging in the first place, so I'm confused about the attempt to still draw boundaries that put it within script-opt territory.  Can't we simply add methods to Expr.h/Stmt.h in this instance ?

* Both `Expr` and `Stmt` are in the `detail` namespace now and that's probably a sufficient way to communicate an interface (in)stability disclaimer to users.  Also not to conflate too much with release timelines, but certainly after Zeek 4.0, there's an awareness of this and we might be free to do more significant changes/re-org to `Expr` / `Stmt` if it helps solve problems in whatever way is most efficient rather than potentially being distracted or misled by a pre-orientation towards objects.

Reply:
> I'm not sure it's worthwhile to jump through the script-optimization demarcation hoops for this case

I agree, but right behind it I have the changes for inlining, and that _will_ add methods that are specific to script optimization.

I really don't care about this one either way, was just trying to lay down a framework that would work going forward, given @rsmmr earlier commented that the optimization stuff should be well-delineated.

> * Both `Expr` and `Stmt` are in the `detail` namespace now and that's probably a sufficient way to communicate an interface (in)stability disclaimer to users

Note that there's something like 4,000 LOC coming as additions to `Expr.cc`/`Stmt.cc`, hence the desireability of somehow separating it out ...